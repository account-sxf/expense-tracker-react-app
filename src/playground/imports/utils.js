// export { remove, flatten as default }

/*
  remove([1, 2, 3, 4, 5, 6], [1, 3, 6]); // [2, 4, 5]
*/
export function remove(arr1, arr2) {
  if (!Array.isArray(arr1) || !Array.isArray(arr2)) {
    throw new Error('expected both arguments to be arrays');
  }
  var result = [];
  var len = arr1.length;
  for (var i = 0; i < len; i++) {
    var elem = arr1[i];
    if (arr2.indexOf(elem) == -1) {
      result.push(elem);
    }
  }
  return result;
}

/*
  flatten([[1, [2, 3]], [[4, 5], 6, 7, [8, 9]]]);
  // [1, 2, 3, 4, 5, 6, 7, 8, 9]
*/
export function flattenHelper(arr, depth) {
  var stack = arr.slice();
  var result = [];

  while (stack.length) {
    var item = stack.pop();

    if (Array.isArray(item) && depth > 0) {
      stack.push.apply(stack, item);
      depth--;
    } else {
      result.push(item);
    }
  }

  return result.reverse();
}

function flatten(arr, depth) {
  if (!Array.isArray(arr)) {
    throw new Error('expected an array');
  }

  if (depth !== undefined && typeof depth !== 'number') {
    throw new Error('depth expects a number');
  }

  return flattenHelper(arr, typeof depth === 'number' ? depth : Infinity);
}
export default flatten

/*
  const cubedRoot = partial(Math.pow, _, 1/3);
  cubedRoot(64); // 4
  const getRoot = partial(Math.pow, 64);
  getRoot(1/2); // 8
*/
export function partial(fn /*, arg1, arg2 etc */) {
  var partialArgs = [].slice.call(arguments, 1);

  if (!partialArgs.length) {
    return fn;
  }

  return function () {
    var args = [].slice.call(arguments);
    var derivedArgs = [];

    for (var i = 0; i < partialArgs.length; i++) {
      var thisPartialArg = partialArgs[i];

      derivedArgs[i] =
        thisPartialArg === undefined ? args.shift() : thisPartialArg;
    }

    return fn.apply(this, derivedArgs.concat(args));
  };
}