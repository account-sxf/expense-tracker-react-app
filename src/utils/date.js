const pad = num => num.toString().padStart(2, '0')

export const utcDateOnly = (time) => {
  const dateTime = new Date(time)
  const day = pad(dateTime.getDate())
  const month = pad(dateTime.getMonth() + 1)
  const year = pad(dateTime.getFullYear())
  return `${year}/${month}/${day}`
}

export const utcDateTime = (time) => {
  const dateTime = new Date(time)
  const day = pad(dateTime.getDate())
  const month = pad(dateTime.getMonth() + 1)
  const year = pad(dateTime.getFullYear())
  const hour = pad(dateTime.getHours())
  const mins = pad(dateTime.getMinutes())
  return `${year}/${month}/${day}T${hour}:${mins}`
}

export const startOfMonth = () => {
  const dateTime = new Date()
  const day = '01'
  const month = pad(dateTime.getMonth() + 1)
  const year = pad(dateTime.getFullYear())
  const startOfMonth = new Date(`${year}-${month}-${day}T00:00:00.000Z`)
  return startOfMonth.getTime()
}

export const endOfMonth = () => {
  const dateTime = new Date()
  const day = '01'
  const month = pad(dateTime.getMonth() + 2)
  const year = pad(dateTime.getFullYear())
  const endOfMonth = new Date(`${year}-${month}-${day}T00:00:00.000Z`)
  return endOfMonth.getTime()
}