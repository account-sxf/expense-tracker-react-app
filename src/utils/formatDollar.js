const nf = new Intl.NumberFormat('en-NZ', {
  style: "currency",
  currency: "NZD",
  maximumFractionDigits: 2,
  roundingIncrement: 1
})

export default (cents) => {
  const dollars = cents / 100
  return nf.format(dollars)
}