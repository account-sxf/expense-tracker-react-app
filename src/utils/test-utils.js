import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import createStore from '../store'
import { BrowserRouter } from 'react-router-dom'
import expenses from '../tests/fixtures/expenses'

export function renderWithWrappers(
  ui, {
    // get test data
    preloadedState = { expenses: { items: [...expenses] } },
    // create store with the preloaded test data
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {

  // Wrapper for react router and redux for integrated tests
  function Wrapper({ children }) {
    return (
      <BrowserRouter>
        <Provider store={store}>
          {children}
        </Provider>
      </BrowserRouter>
    )
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export function renderWithStoreOnly(
  ui, {
    // get test data
    preloadedState = { expenses: { items: [...expenses] } },
    // create store with the preloaded test data
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {

  // Wrapper for react router and redux for integrated tests
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    )
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}