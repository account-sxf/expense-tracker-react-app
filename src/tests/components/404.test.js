/**
 * @jest-environment jsdom
 */
import React from "react"
import '@testing-library/jest-dom'
import NotFoundPage from "../../components/404"
import { renderWithWrappers } from '../../utils/test-utils'

test("Test 404 page", () => {
  const { asFragment } = renderWithWrappers(<NotFoundPage />)
  expect(asFragment()).toMatchSnapshot()
})
