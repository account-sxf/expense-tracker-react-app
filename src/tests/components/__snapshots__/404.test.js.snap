// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 404 page 1`] = `
<DocumentFragment>
  <div>
    404! - 
    <a
      href="/"
    >
      Dashboard
    </a>
  </div>
</DocumentFragment>
`;
